knative-operator/pkg/common/telemetry/telemetry_controller.go

```diff
@@ -0,0 +1,122 @@
+package telemetry
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/runtime"
+	kafkasourcev1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	eventingsourcesv1beta1 "knative.dev/eventing/pkg/apis/sources/v1beta1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type op int
+
+const (
+	inc op = iota
+	dec
+)
+
+var typesAndMetrics = []runtime.Object{
+	&servingv1.Service{},
+	&servingv1.Revision{},
+	&servingv1.Route{},
+	&eventingsourcesv1beta1.PingSource{},
+	&eventingsourcesv1beta1.ApiServerSource{},
+}
+
+// Add creates a new Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileResourcesForTelemetry{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("telemetry-resources-discovery-controller", mgr, controller.Options{Reconciler: r})
```

Comment:
What happens here if we don't provide a Reconciler in the options?

Reply:
nvm, throws an error.

Reply:
Let me check to see how it will behave.

Reply:
@markusthoemmes the most limited example I see is [this](https://github.com/kubernetes-sigs/controller-runtime/blob/af24f3b6f46793c3210224bdd00c2593970942f1/pkg/controller/example_test.go#L44-L48) one.

Reply:
Nice, yeah let's make use of that!

Reply:
Reconciler [needs](https://github.com/kubernetes-sigs/controller-runtime/blob/af24f3b6f46793c3210224bdd00c2593970942f1/pkg/controller/controller.go#L92) to be specified.